033-OSSDatabaseMigration/Coach/05-online-migration.md

```diff
@@ -0,0 +1,223 @@
+FROM
+    (SELECT DISTINCT
+    tc.table_schema,
+    tc.constraint_name AS foreignkey,
+    tc.table_name,
+    kcu.column_name,
+    ccu.table_schema AS foreign_table_schema,
+    ccu.table_name AS foreign_table_name,
+    ccu.column_name AS foreign_column_name
+    FROM information_schema.table_constraints AS tc
+    JOIN information_schema.key_column_usage AS kcu ON tc.constraint_name = kcu.constraint_name AND tc.table_schema = kcu.table_schema
+    JOIN information_schema.constraint_column_usage AS ccu ON ccu.constraint_name = tc.constraint_name AND ccu.table_schema = tc.table_schema
+WHERE constraint_type = 'FOREIGN KEY'
+    ) S
+    GROUP BY S.table_schema, S.foreignkey, S.table_name, S.foreign_table_schema, S.foreign_table_name
+    ) Q
+    GROUP BY Q.table_schema, Q.table_name;
+```
+
+
+## Steps -- MySQL
+
+Run the export
+
+```bash
+
+ mysqldump -h <container ip> -u contosoapp -p --set-gtid-purged=off --databases wth --no-data  --skip-column-statistics >dump_nodata.sql
+
+```
+
+Import with no data. Create the wth database
+
+```bash
+
+mysql -h mytarget2.mysql.database.azure.com -u contosoapp@mytarget2 -p wth <dump_nodata.sql
+
+```
+
+Run script on the target to drop all the foreign keys
+
+```bash
+drop_fk_query.sql
+```
+
+```sql
+ SET group_concat_max_len = 8192;
+ SELECT GROUP_CONCAT(DropQuery SEPARATOR ';\n') as DropQuery
+    FROM
+    (SELECT
+            KCU.REFERENCED_TABLE_SCHEMA as SchemaName,
+            KCU.TABLE_NAME,
+            KCU.COLUMN_NAME,
+            CONCAT('ALTER TABLE ', KCU.TABLE_NAME, ' DROP FOREIGN KEY ', KCU.CONSTRAINT_NAME) AS DropQuery
+            FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE KCU, information_schema.REFERENTIAL_CONSTRAINTS RC
+            WHERE
+              KCU.CONSTRAINT_NAME = RC.CONSTRAINT_NAME
+              AND KCU.REFERENCED_TABLE_SCHEMA = RC.UNIQUE_CONSTRAINT_SCHEMA
+          AND KCU.REFERENCED_TABLE_SCHEMA = 'wth') Queries
+  GROUP BY SchemaName
+
+```
+
+  Run the script and store the output to a file. Running this the following way requires to format the file again
+
+```bash
+
+ mysql -h mytarget2.mysql.database.azure.com -u contosoapp@mytarget2 -pOCPHack8 wth <drop_fk_query.sql >drop_fk.sql
+ cat drop_fk.sql | sed 's/\\n/\n/g' | sed '/DropQuery/d' >drop.sql
+
+ ```
+ Run the sql script to drop all the FKs
+
+ ```sql
+
+ source drop.sql
+
+ ```
+
+ Grant replication client and save to contosoapp user for DMS online migration
+
+ ```sql
+
+  grant replication slave on *.* to 'contosoapp' ;
+  grant replication client on *.* to 'contosoapp' ;
+
+ ```
+
+ If during the DMS online copy the copy fails with some error message "Migrating data to a mysql other than Azure DB for MySQL is not supported", it is because the
+ user connecting to the target database does not have enough privilege on MySQL. The exact and minimum set of privileges is TBD but this works ( screenshot from Mysql workbench ) See this **[Image](./workbench.png)**
```

Comment:
Change the image link to instead show the image inline at the bottom of this file. It is much more readable that way, no need for an external link.

Reply:
Fixed